### ISSUE TITLE
Custom provisioning

### ISSUE THREAD

**Issue by Swader** [hashicorp/otto#174]
As per [this topic](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/otto-tool/F4zGag8V7l0/W5xoUzgZBgAJ), there should be a way to custom provision an environment.

A lot of extensions need to be manually installed / built, like [this one](https://github.com/nrk/phpiredis) or [this one](http://phalconphp.com), so specifying those as a customization option is absolutely essential.

One idea is specifying dependencies like so:

```
application {
  name = "some-php-app-with-deps"
  type = "php"

  dependency {
    source = "https://github.com/nrk/phpiredis"
    run = ["phpize && ./configure --enable-phpiredis", "make && make install"]
  }
}
```

or something of the sort, or just defining a list of provisioning scripts to execute a-la Vagrant, though that removes a lot of Otto's simplicity.

**Comment by iJackUA** [comment]
Maybe it should have support of Heroku Buildpacks https://devcenter.heroku.com/articles/buildpacks  like [deis.io](http://deis.io/) has. Or some similar "plug&play" concept ?

And nevertheless this approach like you mentioned (a set of shell commands) is "reinventing the wheel" like I have mentioned here https://github.com/hashicorp/otto/issues/151#issuecomment-144702133
Vagrant is not forcing any provisioning - that's why it has very wide adoption.

**Comment by jrnt30** [comment]
I agree @iJackUA. In the keynote Mitchell did mention that they were looking to support the buildpacks in the near future, but I am not sure if that was specific to the Nomad deployment target or not.  

It would be nice to get some perspective on this before pursuing in great depth something like PR #137

**Comment by justincampbell** [comment]
@jrnt30 Unfortunately, Heroku buildpacks would not solve the problem of installing packages. Heroku provides [a large set of packages](https://devcenter.heroku.com/articles/cedar-ubuntu-packages) that are available to your apps. We could install all of these, but I'd like to see how long that takes.

Either way, we would like to avoid users having to specify package names to get their app working, as that takes much of the magic away from Otto, and with too much required configuration we've only added another layer of indirection.

**Comment by themccallister** [comment]
Why not follow suite with the Heroku PHP extension, where you can define, in the Appfile, the custom extension(s) required for the app?

```
application {
  name = "example-otto-php-app"
  type = "php"
}

customization "php" {
    php_version = "5.6.13"
    php_extension = "redis"
    php_extension = "memcached"
}
```

Heroku has more details here: 
https://devcenter.heroku.com/articles/php-support#using-optional-extensions

**Comment by ilyar** [comment]
I agree @themccallister  and would add a propose customization process for `type = "php"`, resolve custom extension and php version:

1. custom form `composer.json`
2. custom form `Appfile`

**Comment by XedinUnknown** [comment]
+1. What's the situation? Is this being considered?

**Comment by Swader** [comment]
@XedinUnknown is correct: http://www.sitepoint.com/use-phpbrew-virtphp

**Comment by XedinUnknown** [comment]
Of course, the range seems to be [limited](https://github.com/phpbrew/phpbrew/issues/492), although I didn't expect that much. Still, it's an elegant solution, I think.
